.github/workflows/test.yml

```diff
@@ -64,13 +69,22 @@ jobs:
         with:
           python-version: '3.8'
 
-      - name: Setup Python package dependencies
+      - name: Cache pip
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-python-${{ hashFiles('**/*requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-python-${{ hashFiles('**/*requirements.txt') }}
+            ${{ runner.os }}-python-
+
+      - name: Update pip
         run: |
-            # Install development requirements, package requirements, and frozen image
-            # requirements, with an increasing priority.
-            pip install -r dev-requirements.txt
-            pip install .
-            pip install -r helm-chart/images/binderhub/requirements.txt
+            pip install --upgrade pip
+            pip install --upgrade setuptools wheel
+
+      - name: Setup Python package dependencies
+        run: pip install -r dev-requirements.txt -r helm-chart/images/binderhub/requirements.txt .
```

Comment:
Well, you're installing some set of dependencies, then other one with more strict checks and then another one there all the versions are fixed. This leads to downloading, installing and then uninstalling lots of dependencies' versions which is completely useless:
```
...
 Installing collected packages: six, rsa, protobuf, cffi, cachetools, pytz, markupsafe, googleapis-common-protos, google-auth, cryptography, attrs, sqlalchemy, ruamel.yaml, grpcio, google-api-core, websocket-client, prometheus-client, jinja2, google-cloud-core, alembic, kubernetes, jupyterhub, google-cloud-logging
  Attempting uninstall: six
    Found existing installation: six 1.16.0
    Uninstalling six-1.16.0:
      Successfully uninstalled six-1.16.0
  Attempting uninstall: rsa
    Found existing installation: rsa 4.7.2
    Uninstalling rsa-4.7.2:
      Successfully uninstalled rsa-4.7.2
  Attempting uninstall: cffi
    Found existing installation: cffi 1.14.5
    Uninstalling cffi-1.14.5:
      Successfully uninstalled cffi-1.14.5
  Attempting uninstall: cachetools
    Found existing installation: cachetools 4.2.2
    Uninstalling cachetools-4.2.2:
      Successfully uninstalled cachetools-4.2.2
  Attempting uninstall: markupsafe
    Found existing installation: MarkupSafe 2.0.0
    Uninstalling MarkupSafe-2.0.0:
      Successfully uninstalled MarkupSafe-2.0.0
  Attempting uninstall: google-auth
    Found existing installation: google-auth 1.30.0
    Uninstalling google-auth-1.30.0:
      Successfully uninstalled google-auth-1.30.0
  Attempting uninstall: cryptography
    Found existing installation: cryptography 3.4.7
    Uninstalling cryptography-3.4.7:
      Successfully uninstalled cryptography-3.4.7
  Attempting uninstall: attrs
    Found existing installation: attrs 21.2.0
    Uninstalling attrs-21.2.0:
      Successfully uninstalled attrs-21.2.0
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 1.4.15
    Uninstalling SQLAlchemy-1.4.15:
      Successfully uninstalled SQLAlchemy-1.4.15
  Attempting uninstall: ruamel.yaml
    Found existing installation: ruamel.yaml 0.17.4
    Uninstalling ruamel.yaml-0.17.4:
      Successfully uninstalled ruamel.yaml-0.17.4
  Attempting uninstall: websocket-client
    Found existing installation: websocket-client 0.59.0
    Uninstalling websocket-client-0.59.0:
      Successfully uninstalled websocket-client-0.59.0
  Attempting uninstall: prometheus-client
    Found existing installation: prometheus-client 0.10.1
    Uninstalling prometheus-client-0.10.1:
      Successfully uninstalled prometheus-client-0.10.1
  Attempting uninstall: jinja2
    Found existing installation: Jinja2 3.0.0
    Uninstalling Jinja2-3.0.0:
      Successfully uninstalled Jinja2-3.0.0
  Attempting uninstall: alembic
    Found existing installation: alembic 1.6.2
    Uninstalling alembic-1.6.2:
      Successfully uninstalled alembic-1.6.2
  Attempting uninstall: kubernetes
    Found existing installation: kubernetes 17.17.0
    Uninstalling kubernetes-17.17.0:
      Successfully uninstalled kubernetes-17.17.0
  Attempting uninstall: jupyterhub
    Found existing installation: jupyterhub 1.4.1
    Uninstalling jupyterhub-1.4.1:
      Successfully uninstalled jupyterhub-1.4.1
Successfully installed alembic-1.5.4 attrs-20.3.0 cachetools-4.2.1 cffi-1.14.4 cryptography-3.4.4 google-api-core-1.26.0 google-auth-1.25.0 google-cloud-core-1.6.0 google-cloud-logging-1.15.1 googleapis-common-protos-1.52.0 grpcio-1.35.0 jinja2-2.11.3 jupyterhub-1.2.2 kubernetes-9.0.1 markupsafe-1.1.1 prometheus-client-0.9.0 protobuf-3.14.0 pytz-2021.1 rsa-4.7 ruamel.yaml-0.16.12 six-1.15.0 sqlalchemy-1.3.23 websocket-client-0.57.0
```

Instead you can pass all the requirements in the one line. Pip will check the entire set of constraints and then fetch the most suitable ones. That should decrease installation time.

Also pip can detect incompatibility of dependencies' versions. But this can cause CI/CD failure only is you're installing all the dependencies in the same pip run.
If you installed some dependency in one pip call and then installed another dependency which is not compatible with a previous one in another pip call, you'll see a warning instead of an error.
IMHO it's better to see an error in just your CI/CD workflow before the release than receiving a lot of issues after publishing a broken release.